### DISCUSSION TITLE
FFmpeg in .NET MAUI

### DISCUSSION THREAD

**Discussion by snpkhy** [dotnet/maui#26670] [Q&A]
Is it possible to use FFmpeg in .NET MAUI? Personally I have tried to use .so files, but not all are working correct... Also tried to build FFmpegkit, for linux it has built, for android no.. FFmpeg.Autogen I have also used. In WPF it worked. I also tried to write ffmpeg code in console program and using SignalR to communicate with MAUI project. In android nothing has worked... This message is my last hope! 
PS: VLC and MediaElement do not recommend please, cause I wanna to use unmaneged code. Also it is not good looking and in MediaElemnt i can not render rtsp. For me it is the most important to recieve and decode frames from camera in live mode.

**Comment by AndrewFrancisSE** [comment]
I wanted to ask if you've managed to get it working, particularly on iOS? I'm facing similar issues and would really appreciate it if you could share any progress or experience you've had.

**Reply by snpkhy** [reply]
That's how I made it work. All you need to do is just build the ffmpeg libraries for your operating system. In my case, I had to build not the .dll, but the .so. I would advise you to look for more info on the official ffmpeg website. There may even be already built libraries that you need. Also read about NativeBindings in MAUI. But in general, it's not a problem to add a library type. You just need to put them in Resources and not in the folder that is in the root of the project, but in the folder Platforms and then in the OS you need. In my case, it was the folder Platforms/Android/Resources. Then you just set the appropriate properties for your libraries in the build and pull methods from it. Or write a wrapper right away, as I did. From C++ to C#. You can read about the C++ CLI or how to use C++ code in .NET. Or if you don't have enough methods, just create a static class and use DllImport to extract a couple of methods. Oh, and very important! The desktop and mobile versions of ffmpeg are different. When I was searching for all this, most people wrote about some kind of ffmpeg-kit. I haven't figured it out yet. If you are interested in something else, write to me, I will be happy to share.

**Reply by ghost** [reply]
@vlad0syk, we wanted to use RTSP on iOS, but it doesn't work [out of the box](https://github.com/CommunityToolkit/Maui/discussions/1715).

Another option was to use FFmpeg to connect to the RTSP stream and convert it to MP4. In the end, I didn't try it since FFmpeg is not natively supported in MAUI.

[CoPilot ](https://copilot.microsoft.com/chats/sZeQedbwasPwdothqHhBA) said it is possible and gave me some code example, however I did not test it, maybe somebody can have an idea from this.

```
// FFmpeg arguments:
// -t 10 : Capture for 10 seconds.
// -c copy : Copy streams without re-encoding.
// -movflags frag_keyframe+empty_moov : Make an MP4 streamable by writing headers early.
// -f mp4 pipe:1 : Output as MP4 and write to stdout.
string arguments = $"-i \"{rspCameraUrl}\" -t 10 -c copy -movflags frag_keyframe+empty_moov -f mp4 pipe:1";

ProcessStartInfo psi = new()
{
    FileName = "ffmpeg", // Ensure ffmpeg is in your PATH or specify the full path.
    Arguments = arguments,
    RedirectStandardOutput = true,
    RedirectStandardError = true, // Optional: for debugging FFmpeg logs.
    UseShellExecute = false,
    CreateNoWindow = true
};

byte[] clipBytes;
using (Process ffmpegProcess = new Process { StartInfo = psi })
{
    // Start the ffmpeg process
    ffmpegProcess.Start();

    using (MemoryStream ms = new())
    {
        // Read the binary output (the MP4 stream) from FFmpeg
        await ffmpegProcess.StandardOutput.BaseStream.CopyToAsync(ms);

        // Optionally, you can also read and log error output:
        string errors = await ffmpegProcess.StandardError.ReadToEndAsync();
        if (!string.IsNullOrWhiteSpace(errors))
        {
            Console.WriteLine("FFmpeg errors/warnings:");
            Console.WriteLine(errors);
        }

        // Convert the MemoryStream to a byte array
        clipBytes = ms.ToArray();
    }

    ffmpegProcess.WaitForExit();
}

// Is video there?
if (clipBytes.Length > 0)
{



}
```

**Reply by snpkhy** [reply]
Yeah. Copilot is right. It is a very simple operation with ffmpeg and it is possible.

**Reply by ghost** [reply]
@vlad0syk this thread is set as "unanswered". Should we join together to make an example?

**Reply by snpkhy** [reply]
I can only show on Android, because I created ffmpeg .so files for it and don't have any special hardware for it.

**Reply by snpkhy** [reply]
All you need it is to build ffmpeg libs for IOS.

**Reply by snpkhy** [reply]
Oh and I advise you to pay attention to FFmpeg.Autogen.

**Reply by ghost** [reply]
Can you share more details about FFmpeg.Autogen and building ffmpeg libs for Android?
I am newbe in this area.

**Reply by snpkhy** [reply]
Yeah :) 
Links for already built libs for all platforms - https://ffmpeg.org/download.html#get-sources
A link where is a doc for ffmpeg - https://ffmpeg.org/
A link to FFmpeg.Autogen GitHub - https://github.com/Ruslan-B/FFmpeg.AutoGen
If you need more info - feel free to ask me. I will answer and try to help.

**Reply by ghost** [reply]
Thank you @vlad0syk . 
This is currently outside of my knowldege scope. Trying to understand with CoPilot AI.

Using FFmpeg.AutoGen, we build ffmpeg to iOS, include it in MAUI project and then call it. [Correct](https://copilot.microsoft.com/shares/N1fihZ14Kqo46rXLTL5Fx)?

How did you call FFmpeg.Autogen library in Android?

**Reply by snpkhy** [reply]
Yeah. You build ffmpeg libs and then paste them and call functions from them. To call them I install FFmpeg.Autogen for possibility to use all structures and classes. 

![image](https://github.com/user-attachments/assets/f5f392ab-df06-42d9-8444-74a69a8063e6)

Methods I wrote in one class NativeFFmpeg.cs: 

```
using FFmpeg.AutoGen;
using System.Runtime.InteropServices;

namespace MauiApp1.FFmpeg;

public static unsafe class NativeFFmpeg
{
    [DllImport("libavformat.so")]
    public static extern AVFormatContext* avformat_alloc_context();
    [DllImport("libavutil.so")]
    public static extern int av_dict_set(AVDictionary** opt, string key, string value, int flags);
    [DllImport("libavformat.so")]
    public static extern int avformat_open_input(AVFormatContext** ps, string url, AVInputFormat* fmt, AVDictionary** options);
    [DllImport("libavformat.so")]
    public static extern int avformat_find_stream_info(AVFormatContext* ic, AVDictionary** options);
    [DllImport("libavcodec.so")]
    public static extern AVPacket* av_packet_alloc();
    [DllImport("libavformat.so")]
    public static extern int av_find_best_stream(AVFormatContext* ic, AVMediaType type, int wanted_stream_db, int related_stream, AVCodec** decoder_ret, int flags);
    [DllImport("libavcodec.so")]
    public static extern int av_packet_unref(AVPacket* pkt);
    [DllImport("libavformat.so")]
    public static extern int av_read_frame(AVFormatContext* s, AVPacket* pkt);
    [DllImport("libavcodec.so")]
    public static extern AVPacket* av_packet_clone(AVPacket* src);
    [DllImport("libavutil.so")]
    public static extern int av_strerror(int errnum, byte* errbuf, ulong bufferSize);
    [DllImport("libavcodec.so")]
    public static extern void av_packet_free(AVPacket** pkt);
    [DllImport("libavformat.so")]
    public static extern void avformat_close_input(AVFormatContext** s);
    [DllImport("libavutil.so")]
    public static extern AVFrame* av_frame_alloc();
    [DllImport("libacodec.so")]
    public static extern AVCodec* avcodec_find_decoder(AVCodecID id);
    [DllImport("libavcodec.so")]
    public static extern AVCodecContext* avcodec_alloc_context3(AVCodec* codec);
    [DllImport("libavutill.so")]
    public static extern int av_hwdevice_ctx_create(AVBufferRef** device_ctx, AVHWDeviceType type, string device, AVDictionary* opts, int flags);
    [DllImport("libavcodec.so")]
    public static extern int avcodec_parameters_to_context(AVCodecContext* codec, AVCodecParameters* par);
    [DllImport("libavcodec.so")]
    public static extern int avcodec_open2(AVCodecContext* avctx, AVCodec* codec, AVDictionary** options);
    [DllImport("libavcodec.so")]
    public static extern string avcodec_get_name(AVCodecID id);
    [DllImport("libacodec.so")]
    public static extern void av_frame_free(AVFrame** frame);
    [DllImport("libavcodec.so")]
    public static extern void avcodec_free_context(AVCodecContext** avctx);
    [DllImport("libavutil.so")]
    public static extern void av_buffer_unref(AVBufferRef** buf);
    [DllImport("libavcodec.so")]
    public static extern int avcodec_send_packet(AVCodecContext* avctx, AVPacket* pkt);
    [DllImport("libavcodec.so")]
    public static extern int avcodec_receive_frame(AVCodecContext* avctx, AVFrame* frame);
    [DllImport("libavutil.so")]
    public static extern int av_hwframe_transfer_data(AVFrame* dst, AVFrame* src, int flags);
    [DllImport("libavutil.so")]
    public static extern AVFrame* av_frame_clone(AVFrame* src);
    [DllImport("libavutil.so")]
    public static extern void av_frame_unref(AVFrame* frame);
    [DllImport("libavcodec.so")]
    public static extern int avcodec_parameters_from_context(AVCodecParameters* par, AVCodecContext* codec);
    [DllImport("libavutil.so")]
    public static extern IntPtr av_version_info();
    [DllImport("libavcodec.so")]
    public static extern AVCodecParameters* avcodec_parameters_alloc();
    [DllImport("libswscale.so")]
    public static extern SwsContext* sws_getContext(int srcW, int srcH, AVPixelFormat srcFormat, int dstW, int dstH, AVPixelFormat dstFormat, int flags, SwsFilter* srcFilter, SwsFilter* dstFilter, double* param);
    [DllImport("libavutil.so")]
    public static extern int av_image_get_buffer_size(AVPixelFormat pix_fmt, int width, int height, int align);
    [DllImport("libavutil.so")]
    public static extern int av_image_fill_arrays(ref byte_ptrArray4 dst_data, ref int_array4 dst_linesize, byte* src, AVPixelFormat pix_fmt, int width, int height, int align);
    [DllImport("libswscale.so")]
    public static extern int sws_scale(SwsContext* c, byte*[] srcSlice, int[] srcStride, int srcSliceY, int srcSliceH, byte*[] dst, int[] dstStride);
}
```

Properties for each lib:

![image](https://github.com/user-attachments/assets/36a6eae9-8adb-4adb-b95f-bf2d654b155a)

**Reply by snpkhy** [reply]
It's not really very well done, but it's the only way I've managed to get it up and running.

**Reply by snpkhy** [reply]
Oh and try to use Claude or Gemini. It is better. Also there is very cool feature called Cursor :). Use it all wisely!

**Reply by ghost** [reply]
OK thank you @vlad0syk . I think you made somebody life easier. For me, it made me understand the underlaying platform and connecting to native functions better. For now, I will stop here. Thank you @vlad0syk .

**Comment by landon-codes** [comment]
For desktop development you could bundle binaries for whatever operating system you are building for.

I made a Python script for Windows to automate some of my tasks; all I had to do was use the subprocesses module and call the program.

I would definitely take more effort to implement that for each system, and it likely would not support mobile at all. Maybe you could bundle a lightweight terminal like what Termux did for Android/IOS? That would help with writing less code for each OS, and you would only need to install Ffmpeg for one OS.
